src/librustc/ty/mod.rs

```diff
@@ -2887,6 +2892,13 @@ fn param_env<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     traits::normalize_param_env_or_error(tcx, def_id, unnormalized_env, cause)
 }
 
+pub fn is_trait<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId) -> bool {
+    match tcx.def_key(def_id).disambiguated_data.data {
+        DefPathData::Trait(_) => true,
+        _ => false,
+    }
+}
```

Comment:
Why is this a free function and not a method on `tcx`? Also it's a bit sad to see such a specific hack but I don't have any good suggestion.

Reply:
Good question, should be a method on tcx I agree. By hack you are referring to the whole param-env change, right? (Not the fn itself)

I'm somewhat torn, as I wrote elsewhere -- in a way, I think adding things to param-env is the *right* behavior -- or maybe the answer is that `predices_of` on traits should include `Self: Trait` but WF should use `predicates_defined_on`.